src/tqec/computation/block_graph.py

```diff
@@ -641,7 +641,92 @@ def rotate(
                 pos_map[pipe.v.position],
                 cast(PipeKind, rotated_kind),
             )
-        return rotated
+        return rotated.fix_shadowed_faces()
+
+    def fix_shadowed_faces(self) -> BlockGraph:
+        """Fix the basis of those shadowed faces of the cubes in the graph.
+
+        A pair of face can be shadowed if the cube is connected to two pipes
+        in the same direction. Though these faces are not visible in the 3D
+        visualization, they are still identified by the cube kind and can
+        affect the circuit compilation.
+
+        The basis of the cube is enforced to match the pipes connected to it.
+        If there is unmatched shadowed faces, we try to fix their basis. This
+        will leave some freedom in constructing the model in SketchUp and make
+        everyones' life easier.
+
+        Additionally, for a spatial pass-through, i.e. a cube only connected to
+        two pipes in the same direction, we will fix the cube kind to not be a
+        spatial cube.
+
+        Note that the fixed graph is still not guaranteed to be valid as some
+        other conditions may be violated. You still need to call
+        :py:meth:`~tqec.computation.block_graph.BlockGraph.validate` to check
+        the validity of the graph.
+
+        Returns:
+            A new graph with the shadowed faces fixed.
+        """
+        fixed_cubes: dict[Cube, Cube] = {}
+        for cube in self.cubes:
+            if not isinstance(cube.kind, ZXCube):
+                continue
+            # Group connected pipes by direction
+            pipes_by_direction: dict[Direction3D, list[Pipe]] = {}
+            for pipe in self.pipes_at(cube.position):
+                pipes_by_direction.setdefault(pipe.direction, []).append(pipe)
+            shadowed_directions = {
+                d for d, ps in pipes_by_direction.items() if len(ps) == 2
+            }
+            if not shadowed_directions:
+                continue
+            # No need to handle the case `len(pipes_by_direction) == 0` as there
+            # is no pipes connected to the cube.
+            # No need to handle the case `len(pipes_by_direction) == 3` as it's
+            # a 3D corner that cannot be a valid structure.
+            # Spatial pass-through, ensure that the cube is not a spatial cube
+            if len(pipes_by_direction) in [0, 3]:
+                continue
+            new_kind = cube.kind
+            for shadowed_direction in shadowed_directions:
+                if (
```

Comment:
Maybe add a one-line comment to explicitly write that this is for a pass-through node?